src/webapi/geolocation.rs

```diff
@@ -0,0 +1,135 @@
+#[derive(Clone, Debug, ReferenceType)]
+#[reference(instance_of = "Object")]
+pub struct Coordinates(Reference);
+
+impl Coordinates {
+    /// Return the latitude for the coordinate.
+    pub fn latitude(&self) -> f64 {
+        js! (
+            return @{self}.latitude;
+            ).try_into()
+            .unwrap()
+    }
+
+    /// Return the longitude for the coordinate.
+    pub fn longitude(&self) -> f64 {
+        js! (
+            return @{self}.longitude;
+            ).try_into()
+            .unwrap()
+    }
+
+    /// Return the latitude of the coordinate.
+    pub fn altitude(&self) -> Option<f64> {
+        js! (
+            return @{self}.altitude;
+            ).try_into()
+            .ok()
+    }
+
+    /// Return the accuracy of the coordinate reading.
+    pub fn accuracy(&self) -> f64 {
+        js! (
+            return @{self}.accuracy;
+            ).try_into()
+            .unwrap()
+    }
+
+    /// Return the accuracy of the coordinate reading.
+    pub fn altitude_accuracy(&self) -> Option<f64> {
+        js! (
+            return @{self}.altitudeAccuracy;
+            ).try_into()
+            .ok()
+    }
+
+    /// Return the heading of the coordinate.
+    pub fn heading(&self) -> Option<f64> {
+        js! (
+            return @{self}.heading;
+            ).try_into()
+            .ok()
+    }
+
+    /// Return the speed of the coordinate.
+    pub fn speed(&self) -> Option<f64> {
+        js! (
+            return @{self}.speed;
+            ).try_into()
+            .ok()
+    }
+}
+
+/// Representation of position information which is coordinates at a given time.
+#[derive(Clone, Debug, ReferenceType)]
+#[reference(instance_of = "Object")]
+pub struct Position(Reference);
+
+impl Position {
+    /// Return the coordinate details for this position.
+    pub fn coords(&self) -> Coordinates {
+        Coordinates(
+            js! (
+            return @{self}.coords;
+            ).try_into()
+                .unwrap(),
+        )
+    }
+
+    /// Return the timestamp for this position recording.
+    pub fn timestamp(&self) -> f64 {
+        js! (
+            return @{self}.timestamp;
+            ).try_into()
+            .unwrap()
+    }
+}
+
+/// Watch identifier.
+#[derive(Debug)]
+pub struct WatchId(Value);
+
+impl ::std::default::Default for WatchId {
+    fn default() -> Self {
+        WatchId(Value::Undefined)
+    }
+}
+
+/// API that provides interface to geographic location information.
+#[derive(Debug)]
+pub struct Geolocation;
+
+use webcore::once::Once;
+impl Geolocation {
+    /// Attempt to get current position and invoke callback on success.
+    pub fn get_current_position<F: FnOnce(Position) + 'static>(callback: F) {
+        js! (
+            navigator.geolocation.getCurrentPosition(@{Once(callback)})
+            );
+    }
+
+    /// Watch for position changes and call function with updates.
+    pub fn watch_position<F: Fn(Position) + 'static>(callback: F) -> WatchId {
```

Comment:
This also needs error handling; the callback should probably be something like `FnMut(Result<Position, PositionError>)`.

Reply:
Having a hard time implementing this change. My design shown below where I go from Javascript context to Rust to Javascript compiles but fails at runtime with cryptic message that identifier p is not known:

```
 /// Attempt to get current position and invoke callback on success.
-pub fn get_current_position<F: FnOnce(Position) + 'static>(callback: F) {
+pub fn get_current_position<F: FnMut(Result<Position, PositionError>) + 'static>(mut callback: F) {
+
     js! (
-        navigator.geolocation.getCurrentPosition(@{Once(callback)})
+        navigator.geolocation.getCurrentPosition(function(p) {
+            @{callback(Ok(Position(js!(p).into_reference().unwrap())))};
+        }, function(err) {
+            @{callback(Err(position_err_code_to_enum(&js!(err.code))))};
+        })
     );
 }
```

Reply:
This is AFAIK triggered only once so it should take `FnOnce` instead of `FnMut`.

You're getting an error message as everything inside `@{...}` is Rust code, so inside of the `@{...}` block you're using the `js!` macro again and referring to the `p` variable which is not defined in Rust.

You could probably do something like this:

```
let wrapper = move |is_error: bool, value: Value| {
    // ... call callback here ...
});
js!(
    var callback = @{Once(wrapper)};
    navigator.geolocation.getCurrentPosition(function(pos) {
        callback(false, pos);
    }, function(err) {
        callback(true, pos);
    });
);
```